### PR TITLE
Select the set of indicators for available variables

### PR DESCRIPTION
**Pull request by vindelico**
### Pull Request Checklist:
- [ ] This PR addresses an already opened issue (for bug fixes / features)
    - This PR fixes #xyz
- [ ] (If applicable) Documentation has been added / updated (for bug fixes / features).
- [x] (If applicable) Tests have been added.
- [x] This PR does not seem to break the templates.
- [x] HISTORY.rst has been updated (with summary of main changes).
  - [x] Link to pull request (:pull:`number`) has been added.

### What kind of change does this PR introduce?

Adds the function `xscen.indicators.select_inds_for_avail_vars` to filter the indicators that can be calculated with the variables available in a `xarray.Dataset`.

### Does this PR introduce a breaking change?

No.

### CODE REVIEW

File: tests/test_indicators.py [hunk 1]
```diff
@@ -150,3 +150,59 @@ def test_as_jul(self, restrict_years):
...
+            "indicators", {i.base: i for i in indicators}, reload=True
+        )
+        inds_for_avail_vars_from_module = xs.indicators.select_inds_for_avail_vars(
+            ds=ds, indicators=module
+        )
+        inds_for_avail_vars_from_list_of_tuples = (
+            xs.indicators.select_inds_for_avail_vars(
+                ds=ds, indicators=[(n, i) for n, i in module.iter_indicators()]
+            )
+        )
```
Comment (RondeauG) [inline tests/test_indicators.py:197]: Just FYI, but this could be a good case for a `@pytest.mark.parametrize`.

File: xscen/indicators.py [hunk 2]
```diff
@@ -291,3 +290,55 @@ def func(ds, *, ind, nout):
 
     func.__name__ = ind.identifier
     return partial(func, ind=ind, nout=nout)
+
+
+def select_inds_for_avail_vars(
+    ds: xr.Dataset,
+    indicators: Union[
+        str, Path, Sequence[Indicator], Sequence[tuple[str, Indicator]], ModuleType
```
Comment (RondeauG) [inline xscen/indicators.py:298]: `os.PathLike` is more robust than just `Path`.
```suggestion
        str, os.PathLike, Sequence[Indicator], Sequence[tuple[str, Indicator]], ModuleType
```

File: xscen/indicators.py [hunk 3]
```diff
@@ -291,3 +290,55 @@ def func(ds, *, ind, nout):
...
+    ],
+) -> Sequence[Indicator]:
+    """Filter the indicators for which the necessary variables are available.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Dataset to use for the indicators.
+    indicators : Union[str, Path, Sequence[Indicator], Sequence[Tuple[str, Indicator]]]
```
Comment (RondeauG) [inline xscen/indicators.py:307]:
```suggestion
    indicators : Union[str, os.PathLike, Sequence[Indicator], Sequence[Tuple[str, Indicator]]]
```

File: xscen/indicators.py [hunk 4]
```diff
@@ -291,3 +290,55 @@ def func(ds, *, ind, nout):
...
     func.__name__ = ind.identifier
     return partial(func, ind=ind, nout=nout)
+
+
+def select_inds_for_avail_vars(
+    ds: xr.Dataset,
+    indicators: Union[
+        str, Path, Sequence[Indicator], Sequence[tuple[str, Indicator]], ModuleType
+    ],
+) -> Sequence[Indicator]:
```
Comment (RondeauG) [inline xscen/indicators.py:300]: I think this was wrong?

```suggestion
) -> ModuleType:
```

File: xscen/indicators.py [hunk 5]
```diff
@@ -291,3 +290,55 @@ def func(ds, *, ind, nout):
...
+
+    Returns
+    -------
+    ModuleType – An indicator module.
+
+    See Also
+    --------
+    xclim.indicators, xclim.core.indicator.build_indicator_module_from_yaml
+    """
+    # transform indicator input into a list of tuples (name, indicator)
```
Comment (RondeauG) [inline xscen/indicators.py:321]:
```suggestion
    # Transform the 'indicators' input into a list of tuples (name, indicator)
```

File: xscen/indicators.py [hunk 6]
```diff
@@ -291,3 +290,55 @@ def func(ds, *, ind, nout):
...
+    is_list_of_tuples = isinstance(indicators, list) and all(
+        isinstance(i, tuple) for i in indicators
+    )
+    if isinstance(indicators, (str, Path)):
```
Comment (RondeauG) [inline xscen/indicators.py:325]:
```suggestion
    if isinstance(indicators, (str, os.PathLike)):
```

File: xscen/indicators.py [hunk 7]
```diff
@@ -291,3 +290,55 @@ def func(ds, *, ind, nout):
...
+    }
+    available_inds = [
+        (name, ind)
+        for var in available_vars
+        for name, ind in indicators
+        if var in ind.parameters.keys()
+    ]
+    return xc.core.indicator.build_indicator_module(
+        "inds_for_avail_vars", available_inds
+    )
```
Comment (RondeauG) [inline xscen/indicators.py:344]: What happens if there is no match?

**Review by RondeauG — APPROVED** [review]
LGTM! The only thing that's maybe missing is a check to handle the cases where no match is found (+associated test).

File: tests/test_indicators.py (hunk 1 above)
Comment (vindelico) [reply tests/test_indicators.py:197]: Done.

File: xscen/indicators.py (hunk 7 above)
Comment (vindelico) [reply xscen/indicators.py:344]: > What happens if there is no match?

A module of len == 0 is returned.
I think that's good with the requirement, test added.
Added to doc: 
```

Returns
---------
ModuleType – An indicator module of length [0, n].
```

File: xscen/indicators.py (hunk 2 above)
Comment (vindelico) [reply xscen/indicators.py]: Fixed.

File: xscen/indicators.py (hunk 3 above)
Comment (vindelico) [reply xscen/indicators.py]: Fixed.

File: xscen/indicators.py [hunk 8]
```diff
@@ -293,3 +292,59 @@ def func(ds, *, ind, nout):
...
+    ],
+) -> ModuleType:
+    """Filter the indicators for which the necessary variables are available.
+
+    Parameters
+    ----------
+    ds : xr.Dataset
+        Dataset to use for the indicators.
+    indicators : Union[str, os.PathLike, Sequence[Indicator], Sequence[Tuple[str, Indicator]]]
+        Path to a YAML file that instructs on how to calculate missing variables.
```
Comment (juliettelavoie) [inline xscen/indicators.py:314]: missing variables ? I think this probably is a copy-paste from somewhere else, but it is a bit confusing here as in this case the variables are the input of the indicators (defined in the yaml)

**Review by juliettelavoie — APPROVED** [review]

File: xscen/indicators.py (hunk 8 above)
Comment (vindelico) [reply xscen/indicators.py]: Good catch. Changed.

**Pull request merged**